From 6be5110dd6051225b9fee10b1af506948b082936 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Tue, 15 Jul 2025 17:27:43 +0800
Subject: [PATCH 194/254] mtk: hostapd: fix cca & csa collision issue

When a CCA is triggered during CSA finalization, there is a slim chance
that the CCA beacon will contain CSA IE due to the following sequence of
events.
1. CSA finalization is in progress while a CCA request is triggered
in hostapd.
2. The CCA beacon is built with CSA IE included, since CSA is in progress.
3. CSA finalization completes in kernel, and the CSA finish event is ready to be
sent to hostapd.
4. The CCA request is sent to the kernel and processed without being blocked, as the
csa_active flag is being reset in step 3.

Therefore, delay the CCA request when CSA is in progress.
A similar issue can occur when CSA is triggered during the CCA
finalization.
In this case, simply clean up the CCA parameter, as the CCA will be aborted
by kernel when CSA starts.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>

When CSA is triggered, clear the CCA parameters in hostapd
if CCA is in progress, so that the beacon will not contain
the CCA IE, as the kernel will abort CCA.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 src/ap/hostapd.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 30561db07..d6627a6ed 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -5100,6 +5100,13 @@ static int hostapd_fill_csa_settings(struct hostapd_data *hapd,
 		hostapd_cleanup_cs_params(hapd);
 	}
 
+	/* The color change process will be aborted when a channel switch is started.
+	 * Therefore, clear the CCA parameters here so that the beacon will not
+	 * contain the CCA IE.
+	 */
+	if (hapd->cca_in_progress)
+		hostapd_cleanup_cca_params(hapd);
+
 	switch (settings->freq_params.bandwidth) {
 	case 80:
 		if (settings->freq_params.center_freq2)
@@ -5499,6 +5506,13 @@ static void hostapd_switch_color_timeout_handler(void *eloop_data,
 	unsigned int b;
 	int i, r;
 
+	/* If CSA is in progress, delay the color change */
+	if (hapd->csa_in_progress) {
+		eloop_register_timeout(5, 0, hostapd_switch_color_timeout_handler,
+				       hapd, NULL);
+		return;
+	}
+
 	 /* CCA can be triggered once the handler constantly receives
 	  * color collision events to for at least
 	  * DOT11BSS_COLOR_COLLISION_AP_PERIOD (50 s by default). */
-- 
2.45.2

